ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/io/BigQuery.java

```diff
@@ -0,0 +1,124 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package com.mozilla.telemetry.ingestion.io;
+
+import com.google.cloud.bigquery.BigQueryError;
+import com.google.cloud.bigquery.InsertAllRequest;
+import com.google.cloud.bigquery.InsertAllResponse;
+import com.google.cloud.bigquery.TableId;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+public class BigQuery {
+
+  public static class Write implements Function<Write.TableRow, CompletableFuture<Write.TableRow>> {
+
+    private final com.google.cloud.bigquery.BigQuery bigquery;
+    private final Map<TableId, Batch> batches = new HashMap<>();
+    private final int maxBytes = 10000000; // HTTP request size limit: 10 MB
+    private final int maxMessages = 10000; // Maximum rows per request: 10,000
+    private final long maxDelayMillis = 1000; // default to 1 second
+
+    public Write(com.google.cloud.bigquery.BigQuery bigquery) {
+      this.bigquery = bigquery;
+    }
+
+    @Override
+    public CompletableFuture<TableRow> apply(TableRow row) {
+      return Optional.ofNullable(batches.get(row.tableId)).flatMap(batch -> batch.add(row))
+          .orElseGet(() -> {
+            final Batch batch = new Batch(bigquery, row.tableId);
+            final CompletableFuture<TableRow> response = batch.add(row)
+                .orElseThrow(() -> new IllegalArgumentException("Empty batch rejected message"));
+            batches.put(row.tableId, batch);
+            return response;
+          });
+    }
+
+    /**
+     * POJO of the information necessary to insert a row into BigQuery.
+     */
+    public static class TableRow {
+
+      public final TableId tableId;
+      public final int byteSize;
+      public final Map<String, Object> content;
+
+      public TableRow(TableId tableId, int byteSize, Map<String, Object> content) {
+        this.tableId = tableId;
+        this.byteSize = byteSize;
+        this.content = content;
+      }
+    }
+
+    private class WriteErrors extends RuntimeException {
+
+      public final List<BigQueryError> errors;
+
+      private WriteErrors(List<BigQueryError> errors) {
+        this.errors = errors;
+      }
+    }
+
+    private class Batch {
+
+      private final com.google.cloud.bigquery.BigQuery bigquery;
+      private final InsertAllRequest.Builder builder;
+      private final CompletableFuture<Void> full;
+      private final CompletableFuture<InsertAllResponse> result;
+
+      int size = 0;
+      int byteSize = 0;
```

Comment:
Getting synchronized access right is always a little scary, but it looks like you're doing a reasonable thing here and only accessing or mutating these from synchronized methods.